### PR TITLE
[docs] add note about pyodide support

### PR DESCRIPTION
**Pull request by jameslamb**
This is SO COOL .... I learned tonight that Pyodide is building a version of the `lightgbm` Python package that runs in a web browser.

https://github.com/pyodide/pyodide/tree/main/packages/lightgbm

Try it yourself.... go to https://pyodide.org/en/stable/console.html (recommended from https://pyodide.org/en/stable/usage/index.html), and run the following:

```python
import lightgbm as lgb
from sklearn.datasets import make_regression
from sklearn.metrics import r2_score

X, y = make_regression(n_samples=10_000, n_features=5, n_informative=5)
bst = lgb.train(
    train_set=lgb.Dataset(X, label=y),
    params={"objective": "regression", "verbose": -1}
)

print(r2_score(y, bst.predict(X)))
# 0.995701411473822
```

image

🤯 🤯 🤯 🤯 

This proposes adding a note about that to the README.

### CODE REVIEW

File: README.md
Comment (StrikerRUS) [inline README.md:142]: Refer to #6655.
```suggestion
`pyodide` (run `lightgbm` Python-package in a web browser): https://github.com/pyodide/pyodide
```

**Review by StrikerRUS — APPROVED** [review]
LGTM! Interesting project indeed! I have never heard about it before.

**Comment by agriyakhetarpal** [comment]
Thanks for the note! Would y'all consider helping us where you can by adding a few representatives/core developers/maintainers as maintainers for [LightGBM's recipe](https://github.com/pyodide/pyodide/blob/0ece5258aea9ee6848e7128b59abd759c20a88b8/packages/lightgbm/meta.yaml)?

This way, users would know who to ping in case they face issues with running LightGBM in Pyodide, and we can reach out when we face troubles with updating its version. In case there are concerns about the time required to spend on this – neither of these things should be a frequent occurrence; we usually update versions periodically when we are close to a new release or when we bump the Emscripten version and break the ABI (around twice a year).

**Comment by agriyakhetarpal** [comment]
FYI, I also noted https://github.com/microsoft/LightGBM/issues/5372 – which is a topic that Pyodide resolves by providing LightGBM to run in WASM environments. An out-of-tree CI job for Pyodide as a target could prove to be useful, see: https://pyodide.org/en/latest/development/building-and-testing-packages.html

**Comment by jameslamb** [comment]
> Would y'all consider helping us where you can by adding a few representatives/core developers/maintainers as maintainers for [LightGBM's recipe](https://github.com/pyodide/pyodide/blob/0ece5258aea9ee6848e7128b59abd759c20a88b8/packages/lightgbm/meta.yaml)?

You can add my GitHub handle as a start, I will try to help when I can. Will leave it to other maintainers here to tell you if they'd like to be added.

> *FYI, I also noted https://github.com/microsoft/LightGBM/issues/5372 – which is a topic that Pyodide resolves by providing LightGBM to run in WASM environments.*

Please comment on that issue with links to any relevant examples that someone finding it from search (or subscribed to notifications) could use to understand how to use pyodide to run LightGBM in WASM. Then we can close that on the feature tracker.

> *An out-of-tree CI job for Pyodide as a target could prove to be useful, see: https://pyodide.org/en/latest/development/building-and-testing-packages.html*

Thanks for that link. This is the first time I've heard the phrase "out of tree", and to be honest I don't totally understand what you mean.

Are you asking us to start building WASM wheels and uploading them to https://pypi.org/project/lightgbm/? If so... what's the benefit of doing that when pyodide is already maintaining `lightgbm` builds at https://github.com/pyodide/pyodide/blob/0ece5258aea9ee6848e7128b59abd759c20a88b8/packages/lightgbm/meta.yaml?

**Comment by agriyakhetarpal** [comment]
> You can add my GitHub handle as a start, I will try to help when I can. Will leave it to other maintainers here to tell you if they'd like to be added.

> Please comment on that issue with links to any relevant examples that someone finding it from search (or subscribed to notifications) could use to understand how to use pyodide to run LightGBM in WASM. Then we can close that on the feature tracker.

Thank you for volunteering, @jameslamb – I'll do both of those things!

> > _An out-of-tree CI job for Pyodide as a target could prove to be useful, see: [pyodide.org/en/latest/development/building-and-testing-packages.html](https://pyodide.org/en/latest/development/building-and-testing-packages.html)_
> 
> Thanks for that link. This is the first time I've heard the phrase "out of tree", and to be honest I don't totally understand what you mean.
> 
> Are you asking us to start building WASM wheels and uploading them to [pypi.org/project/lightgbm](https://pypi.org/project/lightgbm/)? If so... what's the benefit of doing that when pyodide is already maintaining `lightgbm` builds at [pyodide/pyodide@`0ece525`/packages/lightgbm/meta.yaml](https://github.com/pyodide/pyodide/blob/0ece5258aea9ee6848e7128b59abd759c20a88b8/packages/lightgbm/meta.yaml)?

No, PyPI does not support uploads WASM wheels yet, unfortunately. That would require a PEP, which I don't think we will be able to find a sponsor for in near time. We currently host WASM wheels for all the packages we are able to build for on our own with an interface via the [JSDelivr CDN](https://www.jsdelivr.com/package/npm/pyodide), thereby providing an entire "distribution" of CPython for WASM rather than just an "implementation".

To answer your question more specifically: yes, we do maintain `lightgbm` builds at the link you've mentioned, but those are tied to a specific version of `lightgbm`, which gets updated manually from time to time (as noted in https://github.com/microsoft/LightGBM/pull/6715#issuecomment-2503837169) and we need to fix any failing tests or deal with changes in public APIs. An "out of tree" (outside the Pyodide repository) CI job in the LightGBM repository means that WASM wheels will be built and tested with PRs and on every commit to the `main` branch, similar to CI jobs for conventional platform targets like Unix/Windows – so that changes that break functionality within a WASM environment are caught and can be potentially addressed as development proceeds. Effectively, that would mean less maintenance for us when we try to update LightGBM's version in-tree (within the Pyodide repository) going forward. :) Also, WebAssembly has 32-bit indexes by default right now, and I see that LightGBM has 32-bit builds already, so my hunch is it shouldn't be too painful to debug issues/failures down the line once a job is set up. There is a chance we might be able to provide a more public and accessible location for hosting wheels on the internet for downstream users (please see https://github.com/pyodide/pyodide/issues/3049) where these WASM wheels can then be uploaded to at a later date.

**Comment by jameslamb** [comment]
Ah great, thanks for that explanation! I'm willing to try this at some point, and we can see how much maintenance effort it is to set up and maintain.

Let's use #5372 to track that work.

I think it won't be a high near-term priority for us, but something I'd like to eventually try adding. Until then, you can `@` me any time in the `pyodide` repo if LightGBM is causing any issues.

> *WebAssembly has 32-bit indexes by default right now, and I see that LightGBM has 32-bit builds already*

Thanks for explicitly calling this out, it might have taken a while for me to discover! We do support a 32-bit build, but only on Windows:

https://github.com/microsoft/LightGBM/blob/6e0b0a8be44b14ade10737288a26aa361a00a18e/build-python.sh#L138-L142

But don't test that configuration in CI.

**Pull request merged**
